README.md
Bodoke CSS
==========

### Still at development.
###### Last release: [0.5.7](https://github.com/dmmarmol/bodokecss/releases/tag/0.5.7)


Bodoke it's a CSS Front-end Framework created with SASS/SCSS, made for supporting multiple grids in a single project with a minimum of custom design.
It's a layout starter project nice and ready for you to design with your own custom identity.

*BodokeCSS* was born as a testing project where i can try any new CSS feature and practice the [CSSOOP](https://www.smashingmagazine.com/2011/12/an-introduction-to-object-oriented-css-oocss/) model.

The base idea of this project is to be responsive and not visualy contaminated.
It supports desktop, mobile and tablet websites using semantic clases in html.

## Support
* IE >= 8
* Firefox >= 30
* Chrome >= 34
* Safari >= 7
* Opera >= 23
* IOS >= 7
* Android >= 4.4
* Blackberry browser >= 10

## How it works?

The `.container` class is the one whe keeps the content centered and makes your site look beautiful. You need to place it right after the `body` tag if you want to prevent the content from touching the viewport.

Then you can put as many `.row`s as you want inside the `.container` and it's mandatory that you put at least one `.col` (or `.column/.columns/.cols`) inside of each one of them.

Set the column width using the column-size classes (eg: `.c6-12` will make that block fill half of the `.row` available space)
``` html
<body>
	<main class="container">
		<div class="row">
			<div class="c6-12 cols">Half row content</div>
			<div class="c6-12 cols">Half row content</div>
		</div>
	</main>
</body>
```

**Changelog:** [Here](https://github.com/dmmarmol/bodokecss/blob/master/CHANGELOG.MD)

Project docs coming soon... But if you are brave enough, you can check the code for yourself (it's self documented!)

## Getting Started

You need to have previously installed [npm](https://docs.npmjs.com/cli/install)

#### Install via npm
```sh
npm install --save bodokecss
```

#### Clone it
Clone the `bodokecss` from your `node_modules` folder into your **assets** or **scss** folder (preferably `scss/`) but it will depend on your project structure

-------------

## Integrate Bodoke into your site

### 1) Link into your html
```html
<head>
	<title>Bodoke Rocks \m/</title>
	<link rel="stylesheet" type="text/css" href="css/bodoke.css">
</head>
<body>
	<!-- What does 'Bodoke' means?, what was this guy thinking? -->
</body>
```

### 2) Include into your sass flow

*Assuming your site structure it's something like this:*
```
/project
|-- *.html
|-- *.php
|-- /node_modules
    |-- bodokecss --> Copy this folder
|-- /src
    |-- css
    |-- js
    |-- images
|-- /scss
    |-- bodokecss --> And paste it here
    |   |-- app.scss
```

Then at the top of your main scss (eg. `app.scss`) file include the next line:

```scss
@import "build/src/bodoke";
```
You can easily customize or expand Bodoke inside `build/src/bodoke.scss` by commenting or uncommenting each `@import` component. *I do not recommend to uncomment the items before 5. (Compnents)*.

-------------

## Compile SCSS files

#### Using Gulp

To make this work, you need to have already installed **[node](http://blog.diegomarmol.com.ar/getting-started-with-node-js-and-gulp-js/)** and **[gulp](http://blog.diegomarmol.com.ar/getting-started-with-node-js-and-gulp-js/)**

If you are a gulp user copy and paste this `task` into your `gulpfile.js` *(see: [blog.diegomarmol.com.ar](http://blog.diegomarmol.com.ar/getting-started-with-node-js-and-gulp-js/))*

After Node was installed, run:
```sh
npm install gulp gulp-sass gulp-autoprefixer --save-dev
```

```js
var gulp 			= require('gulp');
var sass 			= require('gulp-sass');
var autoprefixer 	= require('gulp-autoprefixer');

gulp.task('bdk', function () {
    gulp.src( '.scss/**/*.scss' ) // Set this according the route to your project
		.pipe(sass().on('error', sass.logError))
		.pipe(autoprefixer({
            browsers: AUTOPREFIXER_BROWSERS,
            cascade: false
        }))
		.pipe(gulp.dest( './path/to/your/css/folder' ));
});
```

Now Open the **Terminal** or **CMD**, run `gulp bdk` and your CSS will be up and running!


CHANGELOG.MD
# Changelog

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and [Keep A Changelog](http://keepachangelog.com/CHANGELOG.md).

### Projects involved
* [Tunemypic.com](http://tunemypic.mx/) // V: 0.0.1
* La Caja // V: 0.0.11
* DevBus2014 // V: 0.0.2
* [Youtube/Edu Gadget](http://youtube.com/educacion) // V: 0.0.3
* [Google ContentBox](https://github.com/google/contentbox) // V: 0.0.5
* [Convoy testapp platform](http://audits-platform.appspot.com/) // V: 0.1.0 - New Grid
* DevBus - Egipto // V: 0.1.2 (Amon Carolina)
* Status update // V: 0.1.3 (Amon Demian)
* GCBA - Padres	// V: 0.1.4 (Amon Ed)
* [Codigoquemero](http://codigoquemero.com.ar) // V: 0.1.5 (Amon Fabian)
* [Mundoconsulta](http://mundoconsulta.com)	// V: 0.1.6 (Amon Gabriel)
* [Codigoquemero v2](http://codigoquemero.com.ar) // V: 0.2.2 (Baako Hagen)
* Status Update // V: 0.2.3 (Baako Icarus)
* Status Update // V: 0.2.7 (Baako Maarav)
* Status Update // V: 0.3.0 (Caesar Odysseus)
* Status Update // V: 0.4.0 (Daan Yaffa)
* Status Update // V: 0.5.0 (Eagle Imani)
* [Canvas - Calzado Urbano](http://canvas.com.ar) // V: 0.5.X (Eagle X)
* Status Update // V: 0.5.2 - TAG
* Status Update // V: 0.5.5 - TAG
* Status Update // V: 0.5.52 (Eagle Nadím)
* Status Update // V: 0.5.6
* Status Update // V: X.X.X [unreleased]
	* BodokeCSS is Mobile First

_________________

## [0.5.7 - 2017-08-17]((https://github.com/dmmarmol/bodokecss/releases/tag/0.5.7))
#### Updated
- Remove references in `functions.scss` to variables outside of the file scope

## [0.5.6 - 2017-04-18]((https://github.com/dmmarmol/bodokecss/releases/tag/0.5.6))
#### Added
- Registered *bodokecss* as an npm package (now run `npm install bodokecss --save`)
- Build task with gulp

#### Updated
- Renamed `src/` to `dist/` and `bodokecss/` to `src/`

#### Deleted
- `src/images` dir

## [0.5.X - Changes without version name]
#### Added
- `bodokecss/src/` dir for the assets of your site (eg. Fonts, Background images, etc).
- `bodokecss/lib-extension` for overriding a section of any vendor/lib (eg. `bootstrap-extension.scss`)
- `README.md` files for each dir inside `bodokecss/`
- `@mixin radial-gradient`
- `_media-queries.scss`: New stylesheet with all the neccesary media-query types
- `@respond-to($screen, $limit)`, a Bootstrap `@mixin` for Media Queries managment. Just pass `'xs', 'sm', 'md' or 'lg'` as an argument and all the code inside will have effect inside that resolution

#### Updated
- `@mixin gradient` now uses sass maps as argument, leaving the rest to `gulp-autoprefixer`
- Fixed issue with `@function keyExist()`
- `_text.scss`
- `_buttons.scss` using `@function ifColor()` to determine text color depending on the background
- `@mixin placeholder` renamed to `@mixin bdk-placeholder` (avoiding collision with `bootstrap`)

#### Deleted
- `bodoke-fontastic` font files and scss
- `media/` dir


## [0.5.52 Eagle Nadím](https://github.com/dmmarmol/bodokecss/releases/tag/0.5.51) - 2016-05-29
#### Updated
- `./scss` dir renamed to `./bodokecss`
- `bower.json` fixed typo

## [0.5.5](https://github.com/dmmarmol/bodokecss/releases/tag/0.5.5) - 2016-05-28
#### Added
- `isMap($map)` function. Check if the given object is a [Sass map](https://www.viget.com/articles/sass-maps-are-awesome)

#### Updated
- The colors in `colors.scss` are now Sass maps
- `getColor()` function now search through Sass maps instead of lists

#### Deleted
- `app.scss` and now everything is contained in `bodoke.scss`
- `scss/object` dir
- `src/js/plugins` dir


## [0.5.2](https://github.com/dmmarmol/bodokecss/releases/tag/0.5.2) - 2016-05-9
#### Added
- `@mixin material-shadow($level)` add material design `box-shadow` to the elements. You can pass an `int` for `$level` as an argument `(min: 1, max: 6)` choosing the distance between the element and the bottom (in Z).
- `BodokeCSS` is now a `bower dependecy`
- `@function keyExist` Check if a key is present inside a list and return the key
- `@mixin stop-animation()` and `@mixin infinite-animation()` for animations in `animations.scss`
- `variables.scss`
- `@function rem-to-em` Converts *rem* to *ems*

#### Updated
- `animations.scss` fixed interpolation issue
- `navbar.scss` structure adapted to support Bootstrap 3 `.navbar`

#### Deleted
- Duplicated `_fonts.scss` file
- Duplicated `settings/` dir


## [0.5.0 Eagle Imani](https://github.com/dmmarmol/bodokecss/releases/tag/0.5.0) - 2016-01-21
#### Updated
- Project architecture adapted according to [HugoGiraudel Sass boilerplate template](https://github.com/HugoGiraudel/sass-boilerplate)
- `@mixin animation-delay()` accepts multiple values
- `@function c()` renamed to `getColor()`
- `@mixin delay` renamed to `@mixin transition-delay`
- native `.loading` class renamed using the `$bdk-prefix` -> `.bdk-loading`
- `scss/bodoke/vendor` dir renamed to `scss/bodoke/lib` for avoiding `.gitignore`
- `@mixin bdk-background()` renamed to `background()` because *Bodoke* doesn't need an external mixin library.
- `@mixin animation()` now supports custom `animation-iteration-count` as the second parameter
- `@mixin screen()` has one parameter and is more flexible: The developer can write a list of media-query rules in the CSS way. Eg: `(max-width: 800px)`
- [FontAwesome](http://fontawesome.io) `4.5.0`
- Show/Hide classes for desktop, mobile, tablet and medium screens `@see grid.scss`
- `titles.scss`, `helpers.scss` updated
- Updated **FontAwesome** `v4.45`

#### Added
- [`Gulp.js`](http://gulpjs.com/) for sass compiling
- `@function map-set()`: set new value to a key inside a map.
- `@function map-push()`: adds item to list in map.
- `@function map-deep-get()`: fetch deeply nested values from a map by [HugoGiraudel](https://css-tricks.com/snippets/sass/deep-getset-maps/).
- `@function map-deep-set()`: Set a deeply nested key inside a map.
- `@mixin animation-steps()`: Animate spritesheets with CSS

gulpfile.js
/**
 * gulpfile.js
 * http://localhost:8000/bodokecss
 * livreload script: //localhost:35729/livereload.js?snipver=1
 */

// REQUIRES
var gulp 			= require('gulp');
var sass 			= require('gulp-sass');
var livereload		= require('gulp-livereload');
var autoprefixer 	= require('gulp-autoprefixer');
var gutil 			= require('gulp-util'); // https://www.npmjs.com/package/gulp-util/
var debug 			= require('gulp-debug'); // https://www.npmjs.com/package/gulp-debug/
var del 			= require('del');

var BODOKE 			= './src';
var SRC 			= './dist/';

// ROUTES
var DIR = {
	SASS: 			[ BODOKE+'/**/*.scss' ],
	BODOKEAPP: 		[ BODOKE+'/bodoke.scss' ],
	CSS: 			{
		FOLDER: 	SRC+'css',
		FILES: 		SRC+'css/**/*.css'
	},
	SCRIPTS: {
		FOLDER: 	SRC+'js',
		FILES: 		[SRC+'js/**/*.js', '!'+SRC+'js/**/config.js'],
	},
	FONTS: {
		FOLDER: 	SRC+'fonts',
		FILES: 		SRC+'fonts/**/*'
	},
	HTML: 			[SRC+'**/*.html'],
	PHP: 			[SRC+'**/*.php']
}

var BUILD = {
	MAIN: [].concat(BODOKE, './gulpfile.js'),
	FILES: [].concat(DIR.SASS, DIR.CSS.FILES, DIR.HTML)
}

var AUTOPREFIXER_BROWSERS = [
	'ie >= 8',
	'ie_mob >= 8',
	'ff >= 30',
	'chrome >= 34',
	'safari >= 7',
	'opera >= 23',
	'ios >= 7',
	'android >= 4.4',
	'bb >= 10'
];

// gutil.log( DIR.php );

// TASK
gulp.task('default', ['bdk-scss', 'watch'], function() {

});

// WATCH SCSS TASK
gulp.task('watch', function () {
	livereload.listen();
    gulp.watch( [DIR.HTML, DIR.PHP], ['reload']).on('change', function(file){
    	gulp.src(file.path)
    		.pipe(debug({title: 'Updated:'}));
    });
    gulp.watch( [DIR.SASS], ['bdk-scss']).on('change', function(file){
    	gulp.src(file.path)
    		.pipe(debug({title: 'Updated:'}));
    });
});

// Comple app.scss
gulp.task('bdk-scss', function () {
	gulp.src( DIR.BODOKEAPP )
		.pipe(sass().on('error', sass.logError))
		.pipe(autoprefixer({
            browsers: AUTOPREFIXER_BROWSERS,
            cascade: false
        }))
		.pipe(gulp.dest( DIR.CSS.FOLDER ))
		.pipe(livereload());
});

gulp.task('reload', function() {
	livereload.reload();
});

gulp.task('build', function() {
	del.sync('./build');
	console.log('- REMOVED Build Dir');

	/**
	 * Create the build/
	 */
	gulp.src(BUILD.MAIN).pipe(gulp.dest('./build/'));

	/**
	 * Copy dist/
	 */
	gulp.src(DIR.SCRIPTS.FILES).pipe(gulp.dest('./build/'+SRC+'/js'));
	gulp.src(DIR.FONTS.FILES).pipe(gulp.dest('./build/'+SRC+'/fonts'));
	gulp.src(DIR.CSS.FILES).pipe(gulp.dest('./build/'+SRC+'/css'));

	/**
	 * Copy src/
	 */
	console.log('- CREATED '+BODOKE+' dir');
	gulp.src(BUILD.FILES).pipe(gulp.dest('./build/'+BODOKE));
	console.log('- CREATED '+BUILD.FILES+' dir');

	console.log('- BUILD bodokecss');
});
